% -*- mode: noweb; noweb-default-code-mode: R-mode; -*-
%\VignetteIndexEntry{EnsemblRest primer}
%\VignetteKeywords{}
%\VignetteDepends{}
%\VignettePackage{EnsemblRest}
\documentclass[10pt, a4paper]{article}
\RequirePackage{colortbl}
\RequirePackage{xcolor}
\usepackage{amsmath,pstricks}
\usepackage{hyperref}
\usepackage[hmargin=3cm,vmargin=3.5cm]{geometry}
\usepackage[authoryear,round]{natbib}
\usepackage{fmtcount} % displaying latex counters
\usepackage[T1]{fontenc}
\usepackage[scaled]{helvet}
\usepackage{color}
\usepackage{multirow}
\usepackage{listings}
\usepackage{wrapfig}
\SweaveOpts{keep.source=TRUE}
\definecolor{darkred}{rgb}{0.545,0,0} 
\definecolor{midnightblue}{rgb}{0.098,0.098,0.439} 
\definecolor{palered}{rgb}{1.0,0.5,0.5} 
\definecolor{palegreen}{rgb}{0.5,1.0,0.5} 
\definecolor{silver}{rgb}{0.95,0.95,0.95}
\renewcommand*\familydefault{\sfdefault}
\newcommand{\scscst}{\scriptscriptstyle}
\newcommand{\scst}{\scriptstyle}
\newcommand{\code}[1]{{\ttfamily\texttt{#1}}}
\author{Tim Yates}
\begin{document}
\title{An R interface to the Ensembl REST API}
\maketitle
\tableofcontents
<<echo=FALSE>>=
options(width=80)
@
\section{Introduction}

This package uses the Ensembl REST API\footnote{\url{http://beta.rest.ensembl.org/}} (currently in beta) to extract data from Ensembl into R.

As the REST API is in Beta, this package should also be considered to be in flux and functions/parameters/etc are subject to change as things get finalized.

It could also do with your help.  If you find a problem, something you think could be better, or a better way of doing things, please consider visiting the GitHub project at \url{https://github.com/acbb/EnsemblRest} and posting an issue or a Pull Request.  Thanks!

\section{Available Methods}
To begin with (assuming you have installed this package), you need to load it into your R session:

<<eval=TRUE,echo=TRUE>>=
library( EnsemblRest )
@

The following subsections then list the methods available to you.

\subsection{Information}
\subsubsection{\code{isAlive}}

Firstly, we can check to see that the REST API is accepting calls:
<<eval=TRUE,echo=TRUE>>=
isAlive()
@

\subsubsection{\code{infoSpecies}}

To get a list of available species on the server, you can use the \code{infoSpecies} call\footnote{\url{http://beta.rest.ensembl.org/documentation/info/species}}.

<<eval=TRUE,echo=TRUE>>=
infoSpecies()[1:3] # Just the first 3
@

\subsubsection{\code{infoAssembly}}

The \code{infoAssembly} call\footnote{\url{http://beta.rest.ensembl.org/documentation/info/assembly_info}} returns information about the currently available assemblies in the given species.

<<eval=TRUE,echo=TRUE>>=
infoAssembly( 'human' )
@

\subsubsection{\code{assemblyDetails}}

The \code{assemblyDetails} call\footnote{\url{http://beta.rest.ensembl.org/documentation/info/assembly_stats}} returns information about one of these assemblies.

<<eval=TRUE,echo=TRUE>>=
assemblyDetails( 'X', 'human' )
@

\subsubsection{\code{infoComparas}}

The \code{infoComparas} call\footnote{\url{http://beta.rest.ensembl.org/documentation/info/comparas}} lists the available comparative genomics databases.

<<eval=TRUE,echo=TRUE>>=
infoComparas()
@

\subsubsection{\code{infoData}}

The \code{infoData} call\footnote{\url{http://beta.rest.ensembl.org/documentation/info/data}} shows the data releases available to the REST service

<<eval=TRUE,echo=TRUE>>=
infoData()
@

\subsubsection{\code{infoRest}}

\code{infoRest} shows the current version\footnote{\url{http://beta.rest.ensembl.org/documentation/info/rest}} of the REST service

<<eval=TRUE,echo=TRUE>>=
infoRest()
@

\subsubsection{\code{infoSoftware}}

And finally in the info section, \code{infoSoftware} shows\footnote{\url{http://beta.rest.ensembl.org/documentation/info/software}} the current version of the Ensembl API.

<<eval=TRUE,echo=TRUE>>=
infoSoftware()
@

\subsection{Comparative Genomics}

\subsubsection{\code{geneTree}}

This method\footnote{\url{http://beta.rest.ensembl.org/documentation/info/rest}} fetches the gene tree in New Hampshire format for a given Ensembl gene tree identifier.

<<eval=TRUE,echo=TRUE>>=
geneTree( 'ENSGT00390000003602' )
@

it is also possible to specify the NH format you require (ie: for full format):

<<eval=TRUE,echo=TRUE>>=
geneTree( 'ENSGT00390000003602', nh_format='full' )
@

\subsubsection{\code{homologyById}}

When given an Ensembl Gene ID, returns the homology\footnote{\url{http://beta.rest.ensembl.org/documentation/info/homology_ensemblgene}} information for it.

<<eval=TRUE,echo=TRUE>>=
hResponse = homologyById( 'ENSG00000170037' )
print( hResponse$id ) # The id you queried
print( hResponse$homologies[1:2] ) # Just the top 2 homologies to save room
@

\subsubsection{\code{homologyBySymbol}}

You can also retrieve homology information\footnote{\url{http://beta.rest.ensembl.org/documentation/info/homology_symbol}} given a symbol and a species;

<<eval=TRUE,echo=TRUE>>=
hResponse = homologyBySymbol( 'BRCA2', 'human' )
print( hResponse$id ) # The id you queried
print( hResponse$homologies[1:2] ) # Again, just the top 2 homologies to save room
@

\subsection{Cross References}

Cross references are links to other data about the object of interest.  It should be noted that these other data hold different fields and datatypes, so the results are returned in a \code{data.frame} containing a superset of column names, with non-applicable columns for a given result filled with \code{<NA>}.

As a single object may have multiple synonyms, this will cause the object to exist in multiple rows, one for each synonym. 

\subsubsection{xrefsById}

Firstly we can get all external references\footnote{\url{http://beta.rest.ensembl.org/documentation/info/xref_id}} for a given Ensembl ID:

<<eval=TRUE,echo=TRUE>>=
xrefsById( 'ENSG00000170037' )
@

\subsubsection{xrefsByName}

Or, we can look for an external reference primary accession\footnote{\url{http://beta.rest.ensembl.org/documentation/info/xref_name}} (given a species):

<<eval=TRUE,echo=TRUE>>=
xrefsByName( 'NM_004333', 'human' )
@

\subsubsection{xrefsBySymbol}

And we can finally look up all Ensembl objects referenced by an external symbol for a given species\footnote{\url{http://beta.rest.ensembl.org/documentation/info/xref_external}}:

<<eval=TRUE,echo=TRUE>>=
xrefsBySymbol( 'BRAF', 'human' )
@

\subsection{Lookup}

\subsubsection{lookupId}

<<eval=TRUE,echo=TRUE>>=
lookupId( 'ENSG00000170037' )
@

\subsection{Mapping}

\subsubsection{mapping}

<<eval=TRUE,echo=TRUE>>=
mapping( 'NCBI36', '1..10000:1', 'GRCh37', 'human' )
@

\subsubsection{mappingCdna}

<<eval=TRUE,echo=TRUE>>=
mappingCdna( 'ENST00000288602', '100..300' )
@

\subsubsection{mappingCds}

<<eval=TRUE,echo=TRUE>>=
mappingCds( 'ENST00000288602', '100..300' )
@

\subsubsection{mappingTranslation}

<<eval=TRUE,echo=TRUE>>=
mappingTranslation( 'ENSP00000288602', '100..300' )
@

\subsection{Sequences}

\subsubsection{sequenceById}

<<eval=TRUE,echo=TRUE>>=
sequenceById( 'ENSG00000157764' )
@

\subsubsection{sequenceByRegion}

<<eval=TRUE,echo=TRUE>>=
sequenceByRegion( 'X:1_000_000..1_000_100:1', 'human', format='fasta' )
@

\subsection{Variation}

\subsubsection{variationAllele}

<<eval=TRUE,echo=TRUE>>=
variationAllele( 'C', '9:22125503-22125502:1', 'human' )
@

\subsubsection{variationId}

<<eval=TRUE,echo=TRUE>>=
variationId( 'COSM476', 'human' )
@

\end{document}
